src/components/editor/Bindings/FieldSelection/index.tsx

```diff
@@ -0,0 +1,265 @@
+import {
+    Box,
+    Checkbox,
+    FormControl,
+    FormControlLabel,
+    Stack,
+    Typography,
+} from '@mui/material';
+import {
+    BuiltSpec_Binding,
+    CompositeProjection,
+    ConstraintDictionary,
+    ConstraintTypes,
+    FieldSelectionType,
+    Projection,
+    TranslatedConstraint,
+    ValidationResponse_Binding,
+} from 'components/editor/Bindings/FieldSelection/types';
+import useFieldSelection from 'components/editor/Bindings/FieldSelection/useFieldSelection';
+import {
+    useBindingsEditorStore_recommendFields,
+    useBindingsEditorStore_selectionSaving,
+    useBindingsEditorStore_setRecommendFields,
+    useBindingsEditorStore_setSelectionSaving,
+    useBindingsEditorStore_setSingleSelection,
+} from 'components/editor/Bindings/Store/hooks';
+import { useEditorStore_queryResponse_draftSpecs } from 'components/editor/Store/hooks';
+import ExternalLink from 'components/shared/ExternalLink';
+import FieldSelectionTable from 'components/tables/FieldSelection';
+import { Square } from 'iconoir-react';
+import CheckSquare from 'icons/CheckSquare';
+import { isEqual } from 'lodash';
+import { SyntheticEvent, useCallback, useEffect, useState } from 'react';
+import { FormattedMessage } from 'react-intl';
+import { Schema } from 'types';
+
+interface Props {
+    collectionName: string;
+}
+
+interface FieldMetadata {
+    recommended: boolean;
+    include?: { [field: string]: any };
+    exclude?: string[];
+}
+
+const mapConstraintsToProjections = (
+    projections: Projection[],
+    constraints: ConstraintDictionary,
+    fieldMetadata?: FieldMetadata
+): CompositeProjection[] =>
+    projections.map(({ field, inference, ptr }) => {
+        const constraint: TranslatedConstraint | null = Object.hasOwn(
+            constraints,
+            field
+        )
+            ? {
+                  type: ConstraintTypes[constraints[field].type],
+                  reason: constraints[field].reason,
+              }
+            : null;
+
+        let selectionType: FieldSelectionType | null = 'default';
+
+        if (fieldMetadata) {
+            const { recommended, include, exclude } = fieldMetadata;
+
+            if (include && Object.hasOwn(include, field)) {
+                selectionType = 'include';
+            } else if (exclude?.includes(field)) {
+                selectionType = 'exclude';
+            } else if (!recommended && constraint) {
+                const includeRequired =
+                    constraint.type === ConstraintTypes.FIELD_REQUIRED ||
+                    constraint.type === ConstraintTypes.LOCATION_REQUIRED;
+
+                selectionType = includeRequired ? 'include' : null;
```

Comment:
Any gut feelings on how much this could end up needing to grow? Will we need to keep maintaining this list or does it feel like this will only ever be for required fields?

Reply:
it's core protocol -- hasn't changed in ~two years and hard to change now.